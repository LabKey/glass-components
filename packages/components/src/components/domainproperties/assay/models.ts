/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { IEntityDetails } from '../entities/models';
import { List, Record } from "immutable";
import { SEVERITY_LEVEL_ERROR } from "../constants";
import { DomainDesign } from "../models";
import { Utils } from "@labkey/api";

export class AssayProtocolModel extends Record({
    allowBackgroundUpload: false,
    allowEditableResults: false,
    allowQCStates: false,
    allowSpacesInPath: false,
    allowTransformationScript: false,
    allowPlateMetadata: false,
    autoCopyTargetContainer: undefined,
    autoCopyTargetContainerId: undefined,
    availableDetectionMethods: undefined,
    availableMetadataInputFormats: undefined,
    availablePlateTemplates: undefined,
    backgroundUpload: false,
    description: undefined,
    domains: undefined,
    editableResults: false,
    editableRuns: false,
    exception: undefined,
    metadataInputFormatHelp: undefined,
    moduleTransformScripts: undefined,
    name: undefined,
    plateMetadata: undefined,
    protocolId: undefined,
    protocolParameters: undefined,
    protocolTransformScripts: undefined,
    providerName: undefined,
    saveScriptFiles: false,
    selectedDetectionMethod: undefined,
    selectedMetadataInputFormat: undefined,
    selectedPlateTemplate: undefined,
    qcEnabled: undefined
}) {
    allowBackgroundUpload: boolean;
    allowEditableResults: boolean;
    allowQCStates: boolean;
    allowSpacesInPath: boolean;
    allowTransformationScript: boolean;
    allowPlateMetadata: boolean;
    autoCopyTargetContainer: {};
    autoCopyTargetContainerId: string;
    availableDetectionMethods: [];
    availableMetadataInputFormats: {};
    availablePlateTemplates: [];
    backgroundUpload: boolean;
    description: string;
    domains: List<DomainDesign>;
    editableResults: boolean;
    editableRuns: boolean;
    exception: string;
    metadataInputFormatHelp: any;
    moduleTransformScripts: List<string>;
    name: string;
    plateMetadata: boolean;
    protocolId: number;
    protocolParameters: any;
    protocolTransformScripts: List<string>;
    providerName: string;
    saveScriptFiles: boolean;
    selectedDetectionMethod: string;
    selectedMetadataInputFormat: string;
    selectedPlateTemplate: string;
    qcEnabled: boolean;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    static create(raw: any): AssayProtocolModel {
        let domains = raw.domains || List<DomainDesign>();
        if (raw.domains && Utils.isArray(raw.domains)) {
            domains = List<DomainDesign>(
                raw.domains.map((domain) => {
                    return DomainDesign.create(domain);
                })
            );
        }

        if (raw.protocolTransformScripts && Utils.isArray(raw.protocolTransformScripts)) {
            raw.protocolTransformScripts = List<string>(raw.protocolTransformScripts);
        }
        if (raw.moduleTransformScripts && Utils.isArray(raw.moduleTransformScripts)) {
            raw.moduleTransformScripts = List<string>(raw.moduleTransformScripts);
        }

        // if this is not an existing assay, clear the name property so the user must set it
        const name = !raw.protocolId ? undefined : raw.name;

        // Issue 38685: for new assays, pre-select some required assay properties
        const model = new AssayProtocolModel({...raw, name, domains});
        if (model.isNew()) {
            if (model.allowDetectionMethodSelection() && model.availableDetectionMethods.length > 0) {
                raw.selectedDetectionMethod = List<string>(model.availableDetectionMethods).get(0);
            }
            if (model.allowPlateTemplateSelection() && model.availablePlateTemplates.length > 0) {
                raw.selectedPlateTemplate = List<string>(model.availablePlateTemplates).get(0);
            }
        }

        return new AssayProtocolModel({...raw, name, domains});
    }

    static serialize(model: AssayProtocolModel): any {
        // need to serialize the DomainDesign objects to remove the unrecognized fields
        const domains = model.domains.map((domain) => {
            return DomainDesign.serialize(domain);
        });

        let json = model.merge({domains}).toJS();

        // only need to serialize the id and not the autoCopyTargetContainer object
        delete json.autoCopyTargetContainer;
        delete json.exception;

        return json;
    }

    getDomainByNameSuffix(name: string): DomainDesign {
        if (this.domains.size > 0) {
            return this.domains.find((domain) => {
                return domain.isNameSuffixMatch(name);
            });
        }
    }

    get container() {
        return this.getIn(['domains', 0, 'container']);
    }

    isNew(): boolean {
        return !this.protocolId;
    }

    allowPlateTemplateSelection(): boolean {
        return this.availablePlateTemplates && Utils.isArray(this.availablePlateTemplates);
    }

    allowDetectionMethodSelection(): boolean {
        return this.availableDetectionMethods && Utils.isArray(this.availableDetectionMethods);
    }

    allowMetadataInputFormatSelection(): boolean {
        return this.availableMetadataInputFormats && Utils.isObject(this.availableMetadataInputFormats) && !Utils.isEmptyObj(this.availableMetadataInputFormats);
    }

    validateTransformScripts(): string {
        if (this.protocolTransformScripts === undefined || this.protocolTransformScripts.size === 0) {
            return undefined;
        }

        // make sure we don't have any script inputs that are empty strings
        const hasEmptyScript = this.protocolTransformScripts.some((script, i) => script === undefined || script === null || script.length === 0);
        if (hasEmptyScript) {
            return 'Missing required transform script path.';
        }

        // if not allowSpacesInPath, the path to the script should not contain spaces when the Save Script Data check box is selected
        if (!this.allowSpacesInPath && this.saveScriptFiles) {
            const hasSpacedScript = this.protocolTransformScripts.some((script, i) => script.indexOf(' ') > -1);
            if (hasSpacedScript) {
                return 'The path to the transform script should not contain spaces when the \'Save Script Data for Debugging\' check box is selected.'
            }
        }
    }

    static isValid(model: AssayProtocolModel): boolean {
        return model.hasValidProperties();
    }

    hasValidProperties(): boolean {
        return (this.name !== undefined && this.name !==null && this.name.trim().length > 0)
            && (!this.allowMetadataInputFormatSelection() || Utils.isString(this.selectedMetadataInputFormat))
            && (!this.allowDetectionMethodSelection() || Utils.isString(this.selectedDetectionMethod))
            && (!this.allowPlateTemplateSelection() || Utils.isString(this.selectedPlateTemplate))
            && this.validateTransformScripts() === undefined;
    }
}
