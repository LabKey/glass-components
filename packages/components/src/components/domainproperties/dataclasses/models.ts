/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Draft, immerable, produce } from 'immer';
import { Map, fromJS } from 'immutable';

import { DomainDesign, IDomainField } from '../models';

interface DataClassOptionsConfig {
    category: string;
    description: string;
    name: string;
    nameExpression: string;
    rowId: number;
    sampleSet: number;
}

export interface DataClassModelConfig extends DataClassOptionsConfig {
    domain: DomainDesign;
    exception: string;
}

export class DataClassModel implements DataClassModelConfig {
    [immerable] = true;

    readonly category: string;
    readonly description: string;
    readonly domain: DomainDesign;
    readonly exception: string;
    readonly name: string;
    readonly nameExpression: string;
    readonly rowId: number;
    readonly sampleSet: number;

    constructor(values?: Partial<DataClassModelConfig>) {
        Object.assign(this, values);
    }

    static create(raw: any): DataClassModel {
        let domain = DomainDesign.create({});
        if (raw.domainDesign) {
            domain = DomainDesign.create(raw.domainDesign);
        }

        if (raw.options) {
            const model = new DataClassModel({ ...raw.options, domain });
            return produce(model, (draft: Draft<DataClassModel>) => {
                if (model.category === null) {
                    draft.category = undefined;
                }
                if (model.sampleSet === null) {
                    draft.sampleSet = undefined;
                }
            });
        }

        return new DataClassModel({ ...raw, domain });
    }

    get isNew(): boolean {
        return !this.rowId;
    }

    isValid(defaultNameFieldConfig?: Partial<IDomainField>): boolean {
        return (
            this.hasValidProperties &&
            !this.hasInvalidNameField(defaultNameFieldConfig) &&
            !this.domain.hasInValidFields()
        );
    }

    get hasValidProperties(): boolean {
        return this.name !== undefined && this.name !== null && this.name.trim().length > 0;
    }

    hasInvalidNameField(defaultNameFieldConfig: Partial<IDomainField>): boolean {
        return this.domain && defaultNameFieldConfig ? this.domain.hasInvalidNameField(defaultNameFieldConfig) : false;
    }

    get entityDataMap(): Map<string, any> {
        return fromJS({
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
        });
    }

    get options(): DataClassOptionsConfig {
        return {
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
            category: this.category,
            sampleSet: this.sampleSet,
        };
    }
}
