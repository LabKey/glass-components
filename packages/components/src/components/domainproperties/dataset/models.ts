/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Record} from "immutable";
import {DomainDesign} from "../models";

export interface AdvancedSettingsForm {
    titleColumn?: string;
    discussionSetting?: number;
    fileAttachmentIndex?: boolean;
    entireListIndex?: boolean;
    entireListTitleTemplate?: string;
    entireListIndexSetting?: number;
    entireListBodySetting?: number;
    eachItemIndex?: boolean;
    eachItemTitleTemplate?: string;
    eachItemBodySetting?: number;
    entireListBodyTemplate?: string;
    eachItemBodyTemplate?: string;
}

export class DatasetModel extends Record({
    datasetId: undefined,
    name: undefined,
    typeURI: undefined,
    category: undefined,
    visitDatePropertyName: undefined,
    keyProperty: undefined,
    isDemographicData: undefined,
    label: undefined,
    cohortId: undefined,
    tag: undefined,
    showByDefault: undefined,
    description: undefined,
    sourceAssayName: undefined,
    sourceAssayURL: undefined,
    dataSharing: undefined
}) {
    datasetId: number;
    name: string;
    typeURI?: string;
    category?: string;
    visitDatePropertyName?: string;
    keyProperty?: string;
    isDemographicData?: boolean;
    label?: string;
    cohortId?: number;
    tag?: string;
    showByDefault?: boolean;
    description?: string;
    sourceAssayName?: string;
    sourceAssayURL?: string;
    dataSharing?: string;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    static create(newDataset=null): DatasetModel {
        if (newDataset) {
            let domain = DomainDesign.create(undefined);
            return new DatasetModel({...newDataset, domain});
        }
    }
}
