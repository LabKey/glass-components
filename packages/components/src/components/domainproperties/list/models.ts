/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Record } from "immutable";
import { SEVERITY_LEVEL_ERROR } from "../constants";
import { DomainDesign } from "../models";

export interface AdvancedSettingsForm {
    titleColumn?: string;
    discussionSetting?: number;
    fileAttachmentIndex?: boolean;
    entireListIndex?: boolean;
    entireListTitleTemplate?: string;
    entireListIndexSetting?: number;
    entireListBodySetting?: number;
    eachItemIndex?: boolean;
    eachItemTitleTemplate?: string;
    eachItemBodySetting?: number;
    entireListBodyTemplate?: string;
    eachItemBodyTemplate?: string;
}

export class ListModel extends Record({
    exception: undefined,
    domain: undefined,
    entityId : undefined,
    createdBy : undefined,
    created : undefined,
    modifiedBy : undefined,
    modified : undefined,
    containerId : undefined,
    name : undefined,
    description : undefined,
    lastIndexed : undefined,
    keyName : undefined,
    titleColumn : undefined,
    domainId : undefined,
    keyType : undefined,
    discussionSetting : undefined,
    allowDelete : undefined,
    allowUpload : undefined,
    allowExport : undefined,
    entireListIndex : undefined,
    entireListIndexSetting : undefined,
    entireListTitleSetting : undefined,
    entireListTitleTemplate : undefined,
    entireListBodySetting : undefined,
    entireListBodyTemplate : undefined,
    eachItemIndex : undefined,
    eachItemTitleSetting : undefined,
    eachItemTitleTemplate : undefined,
    eachItemBodySetting : undefined,
    eachItemBodyTemplate : undefined,
    fileAttachmentIndex : undefined,
    listId : undefined,
    entireListTitleSettingEnum : undefined,
    entireListBodySettingEnum : undefined,
    eachItemTitleSettingEnum : undefined,
    eachItemBodySettingEnum : undefined,
    discussionSettingEnum : undefined,
    entireListIndexSettingEnum : undefined,
    containerPath : undefined,
}) {
    exception: string;
    domain: DomainDesign;
    name : string;
    description : string;
    lastIndexed : any; //confirm
    keyName : string;
    titleColumn : null;
    domainId : number;
    keyType : string;
    discussionSetting : number;
    allowDelete : true;
    allowUpload : true;
    allowExport : true;
    entireListIndex : true;
    entireListIndexSetting : number;
    entireListTitleSetting : number;
    entireListTitleTemplate : any; //confirm
    entireListBodySetting : number;
    entireListBodyTemplate : any; //confirm
    eachItemIndex : false;
    eachItemTitleSetting : number;
    eachItemTitleTemplate : any; //confirm
    eachItemBodySetting : number;
    eachItemBodyTemplate : any; //confirm
    fileAttachmentIndex : false;
    listId : number;
    entireListTitleSettingEnum : string;
    entireListBodySettingEnum : string;
    eachItemTitleSettingEnum : string;
    eachItemBodySettingEnum : string;
    discussionSettingEnum : string;
    entireListIndexSettingEnum : string;
    containerPath : string;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    static create(raw: any, defaultSettings=null): ListModel {
        if (defaultSettings) {
            let domain = DomainDesign.create(undefined);
            return new ListModel({...defaultSettings, domain});
        } else {
            let domain = DomainDesign.create(raw.domainDesign);
            return new ListModel({...raw.options, domain});
        }
    }

    getDomainKind(): string {
        if (this.keyType === "Varchar") {
            return "VarList"
        }
        else if (this.keyType === "Integer" || this.keyType === "AutoIncrementInteger") {
            return "IntList"
        }

        return undefined;
    }

    hasValidKeyType(): boolean {
        return this.getDomainKind() !== undefined;
    }

    isNew(): boolean {
        return !this.listId;
    }

    static isValid(model: ListModel): boolean {
        const errDomain = !!model.domain.domainException && model.domain.domainException.severity === SEVERITY_LEVEL_ERROR;
        return !errDomain && model.hasValidProperties() && model.hasValidKeyType();
    }

    hasValidProperties(): boolean {
        return this.name !== undefined && this.name !== null && this.name.trim().length > 0;
    }

    getOptions(): Object {
        let options = this.toJS();
        delete options.exception;
        delete options.domain;
        return options;
    }
}
