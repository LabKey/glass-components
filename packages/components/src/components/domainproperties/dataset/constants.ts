export const DATASET_NAME_TIP = 'This name must be unique. It is used when identifying datasets during data upload.';

export const DATASET_CATEGORY_TIP =
    'Assigning a category to a dataset will group it with similar datasets in the navigator and data browser.';

export const DATASET_LABEL_TIP = 'The name of the dataset shown to users. If no Label is provided, the Name is used.';

export const DATASET_ID_TIP =
    'The unique, numerical identifier for your dataset. It is defined during dataset creation and cannot be modified.';

export const VISIT_DATE_TIP =
    "If the official 'Visit Date' for a visit can come from this dataset, choose the date column to represent it. Note that since datasets can include data from many visits, each visit must also indicate the official 'VisitDate' dataset.";

export const COHORT_TIP = 'Datasets may be cohort specific, or associated with all cohorts.';

export const TAG_TIP = 'Adding a tag provides an additional, flexible way to categorize this dataset.';

export const DATASPACE_TIP =
    "Shared data across studies,   When 'No' is selected (default) each study folder 'owns' its own data rows.  If study has shared visits, then 'Share by Participants' means that data rows are shared across the project, and studies will only see data rows for participants that are part of that study.";

export const DATA_ROW_UNIQUENESS =
    'Choose criteria for how participants and visits are paired with or without an additional data column.';
