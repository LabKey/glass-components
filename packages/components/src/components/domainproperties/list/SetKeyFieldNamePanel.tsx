import React from 'react';
import { Alert, Col, FormControl, Row } from 'react-bootstrap';
import {List} from "immutable";
import {DATE_TYPE, DomainField, IDomainField} from '../models';
import {LabelHelpTip, ListModel} from '../../..';

// TODO: define props - IAppDomainHeader (and some other properties, model, onModelChange, keyField, onAddField)
export class SetKeyFieldNamePanel extends React.PureComponent<any> {

    setKeyField = (fields, newKey) => {
        const newKeyField = fields.get(newKey);
        const updatedNewKeyField = newKeyField.merge({ isPrimaryKey: true, required: true }) as DomainField;
        return fields.set(newKey, updatedNewKeyField) as List<DomainField>;
    };

    unsetKeyField = (fields, prevKey) => {
        const prevKeyField = fields.get(prevKey) as DomainField;
        const updatedPrevKeyField = prevKeyField.merge({isPrimaryKey: false, required: false}) as DomainField;
        return fields.set(prevKey, updatedPrevKeyField) as List<DomainField>;
    };

    addAutoIntField = (onAddField, onModelChange, name, value) => {
        const autoIncrementFieldConfig = {
            required: true,
            name: 'Auto increment key (placeholder)',
            dataType: DATE_TYPE,
            // conceptURI: DATE_TYPE.conceptURI,
            rangeURI: DATE_TYPE.rangeURI,
            isPrimaryKey: true

        } as Partial<IDomainField>;
        onModelChange(name, value);
        onAddField(autoIncrementFieldConfig);
    };

    removeAutoIntField = (fields) => { //TODO RP: propertly identify the field by its dataType
        return fields.filter((field) => {
            return field.get('name') !== 'Auto increment key (placeholder)'
        }) as List<DomainField>;

    };

    onSelectionChange = (e) => {
        const {model, onModelChange, keyField, onAddField, onDomainChange} = this.props;
        const {domain} = model;
        const {fields} = domain;
        const { name, value } = e.target;

        let newFields;

        // Making first selection of key
        if (keyField == '-2') {
            if (value == '-1') {
                this.addAutoIntField(onAddField, onModelChange, name, value); // Selecting auto int key
                return;
            } else {
                newFields = this.setKeyField(fields, value);  // Selecting regular field
            }
        // Changing key from one field to another
        } else {
            if (keyField == '-1') {
                const fieldsNoKey = this.removeAutoIntField(fields); // Auto int to regular field
                newFields = this.setKeyField(fieldsNoKey, value);
            } else if (value == '-1') {
                newFields = this.unsetKeyField(fields, keyField); // Regular to auto int field

                onDomainChange(domain.merge({fields: newFields}));
                console.log("Old key field correct un-set.");

                this.addAutoIntField(onAddField, onModelChange, name, value);
                return;
            } else if (value !== '-1') {
                const fieldsNoKey = this.unsetKeyField(fields, keyField); // Regular to regular field
                newFields = this.setKeyField(fieldsNoKey, value);
            }
        }

        const newKeyField = fields.get(value);
        let keyType;
        if (newKeyField.dataType.name === 'int') {
            keyType = 'Integer';
        } else if (newKeyField.dataType.name === 'string') {
            keyType = 'Varchar';
        }

        const updatedModel = model.merge({
            domain: model.domain.set('fields', newFields),
            keyName: newKeyField.name,
            keyType,
        }) as ListModel;

        onModelChange(name, value, updatedModel);
    };

    render() {
        const { keyField } = this.props;
        let fieldNames = [];
        if (this.props.domain) {
            const fields = this.props.domain.fields;

            fieldNames =
                fields &&
                fields.reduce(function(accum: string[], field: IDomainField) {
                    const dataType = field.dataType.name;
                    if (
                        (dataType == 'string' || dataType == 'int') &&
                        typeof field.name !== 'undefined' &&
                        field.name.trim().length > 0
                    ) {
                        accum.push(field.name);
                    }
                    return accum;
                }, []);
        }
        const autoIntIsPK = (keyField == '-1');
        return (
            <Alert>
                <div>
                    Select a key value for this list which uniquely identifies the item. You can use "Auto integer key"
                    to define your own below.
                </div>
                <Row style={{ marginTop: '15px' }}>
                    <Col xs={3} style={{ color: 'black' }}>
                        Key Field Name
                        <LabelHelpTip
                            title=""
                            body={() => {
                                return <> Only integer or text fields can be made the primary key. </>;
                            }}
                        />
                        *
                    </Col>
                    <Col xs={3}>
                        <FormControl
                            componentClass="select"
                            name="keyField"
                            onChange={e => this.onSelectionChange(e)}
                            value={keyField}
                            style={{ width: '200px' }}
                        >
                            <option disabled value={-2}> Select a field from the list </option>

                            {!autoIntIsPK &&
                                <option value={-1}>Auto integer key</option>
                            }

                            {fieldNames.map((fieldName, index) => {
                                return (
                                    <option value={index} key={index + 1}>
                                        {fieldName}
                                    </option>
                                );
                            })}
                        </FormControl>
                    </Col>
                </Row>
            </Alert>
        );
    }
}
