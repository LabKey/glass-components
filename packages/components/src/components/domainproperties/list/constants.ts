export const DATA_INDEXING_TIP = "Not recommend for large lists with frequent updates, since updating any item will cause re-indexing of the entire list";
export const ENTIRELIST_TITLE_TIP = "Any text you want displayed and indexed as the search result title";
export const EACHITEM_TITLE_TIP = "Example: AutoInt - ${Key} ${value}";
export const CUSTOM_TEMPLATE_TIP = "Example: ${Key} ${value}";
