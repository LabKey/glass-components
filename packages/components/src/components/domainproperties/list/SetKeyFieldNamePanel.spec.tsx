import { mount } from 'enzyme';
import React from 'react';

import renderer from 'react-test-renderer';

import getDomainDetailsJSON from '../../../test/data/list-getDomainDetails.json';
import getDomainDetailsWithAutoIncPKJSON from '../../../test/data/list-getDomainDetails-withAutoIntPK.json';
import { DEFAULT_LIST_SETTINGS } from '../../../test/data/constants';

import { SetKeyFieldNamePanel } from './SetKeyFieldNamePanel';
import { ListModel } from './models';

const emptyNewModel = ListModel.create(null, DEFAULT_LIST_SETTINGS);
const populatedExistingModel = ListModel.create(getDomainDetailsJSON);
const populatedExitingModelWithAutoIncPK = ListModel.create(getDomainDetailsWithAutoIncPKJSON);

describe('SetKeyFieldNamePanel', () => {
    test('new list, default properties', () => {
        const setKeyFieldNamePanel = (
            <SetKeyFieldNamePanel
                model={emptyNewModel}
                onModelChange={jest.fn()}
                domain={emptyNewModel.domain}
                domainIndex={1}
            />
        );

        const tree = renderer.create(setKeyFieldNamePanel).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('existing list, given properties', () => {
        const setKeyFieldNamePanel = (
            <SetKeyFieldNamePanel
                model={populatedExistingModel}
                onModelChange={jest.fn()}
                domain={populatedExistingModel.domain}
                domainIndex={1}
            />
        );

        const tree = renderer.create(setKeyFieldNamePanel).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('list with auto integer key', () => {
        const setKeyFieldNamePanel = (
            <SetKeyFieldNamePanel
                model={populatedExitingModelWithAutoIncPK}
                onModelChange={jest.fn()}
                domain={populatedExistingModel.domain}
                domainIndex={1}
            />
        );

        const tree = renderer.create(setKeyFieldNamePanel).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('key fields that are unnamed, or are not of string or int dataType, are invalid', () => {
        const setKeyFieldNamePanel = mount(
            <SetKeyFieldNamePanel
                model={populatedExistingModel}
                onModelChange={jest.fn()}
                domain={populatedExistingModel.domain}
                domainIndex={1}
            />
        );

        const selectOptionsText = setKeyFieldNamePanel.find('.form-control').text();
        expect(selectOptionsText).toContain('SubjectID');
        expect(selectOptionsText).toContain('Name');
        expect(selectOptionsText).toContain('Family');
        expect(selectOptionsText).toContain('Species');
        expect(selectOptionsText).toContain('MaritalStatus');
        expect(selectOptionsText).toContain('CurrentStatus');
        expect(selectOptionsText).toContain('Gender');

        expect(selectOptionsText).toContain('Auto integer');

        expect(selectOptionsText).not.toContain('Mothers');
        expect(selectOptionsText).not.toContain('Father');
        expect(selectOptionsText).not.toContain('Image');
        expect(selectOptionsText).not.toContain('Occupation');
        expect(selectOptionsText).not.toContain('BirthDate');
        expect(selectOptionsText).not.toContain('CartoonAvailable');
        setKeyFieldNamePanel.unmount();
    });
});
