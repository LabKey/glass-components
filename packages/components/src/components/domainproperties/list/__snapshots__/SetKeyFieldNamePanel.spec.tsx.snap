// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetKeyFieldNamePanel existing list, given properties 1`] = `
<div
  className="alert alert-info"
  role="alert"
>
  <div>
    Select a key value for this list which uniquely identifies the item. You can use "
    Auto integer key
    " to define your own below.
  </div>
  <div
    className="list__set-key-panel row"
  >
    <div
      className="col-xs-3"
    >
      Key Field Name *
      <span
        className="label-help-target"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
          data-icon="question-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <div
      className="col-xs-5"
    >
      <select
        className="form-control"
        name="keyField"
        onChange={[Function]}
        value={0}
      >
        <option
          value={-1}
        >
          Auto integer key
        </option>
        <option
          value={0}
        >
          SubjectID
        </option>
        <option
          value={1}
        >
          Name
        </option>
        <option
          value={2}
        >
          Family
        </option>
        <option
          value={5}
        >
          Species
        </option>
        <option
          value={8}
        >
          MaritalStatus
        </option>
        <option
          value={9}
        >
          CurrentStatus
        </option>
        <option
          value={10}
        >
          Gender
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`SetKeyFieldNamePanel list with auto integer key 1`] = `
<div
  className="alert alert-info"
  role="alert"
>
  <div>
    Select a key value for this list which uniquely identifies the item. You can use "
    Auto integer key
    " to define your own below.
  </div>
  <div
    className="list__set-key-panel row"
  >
    <div
      className="col-xs-3"
    >
      Key Field Name *
      <span
        className="label-help-target"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
          data-icon="question-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <div
      className="col-xs-5"
    >
      <select
        className="form-control"
        name="keyField"
        onChange={[Function]}
        value={0}
      >
        <option
          value={-1}
        >
          Auto integer key
        </option>
        <option
          value={0}
        >
          SubjectID
        </option>
        <option
          value={1}
        >
          Name
        </option>
        <option
          value={2}
        >
          Family
        </option>
        <option
          value={5}
        >
          Species
        </option>
        <option
          value={8}
        >
          MaritalStatus
        </option>
        <option
          value={9}
        >
          CurrentStatus
        </option>
        <option
          value={10}
        >
          Gender
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`SetKeyFieldNamePanel new list, default properties 1`] = `
<div
  className="alert alert-info"
  role="alert"
>
  <div>
    Select a key value for this list which uniquely identifies the item. You can use "
    Auto integer key
    " to define your own below.
  </div>
  <div
    className="list__set-key-panel row"
  >
    <div
      className="col-xs-3"
    >
      Key Field Name *
      <span
        className="label-help-target"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
          data-icon="question-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <div
      className="col-xs-5"
    >
      <select
        className="form-control"
        name="keyField"
        onChange={[Function]}
      >
        <option>
          Select a field from the list
        </option>
        <option
          value={-1}
        >
          Auto integer key
        </option>
      </select>
    </div>
  </div>
</div>
`;
